### ISSUE TITLE
Template does not reflect updated Mirage data

### ISSUE THREAD

**Issue by awrichar** [samselikoff/ember-cli-mirage#1394]
When I update Mirage data during acceptance tests, the HTML document does not reflect the change. For instance, if I add an element with `server.create`, I'd expect the `{{#each}}` block of my template to reflect the newly added element. Am I missing something to trigger this behavior?

I've been all over the documentation of Ember.js and Mirage, and can't figure it out. My application works in the "real world" by updating the store via `DS.store.push()`, and the page reflects the created/updated items immediately. Trying to figure out how to simulate this behavior under Mirage.

```
import { module, test } from 'qunit';
import { visit } from '@ember/test-helpers';
import { setupApplicationTest } from 'ember-qunit';
import setupMirage from 'ember-cli-mirage/test-support/setup-mirage';

module('Acceptance | list widgets', function(hooks) {
  setupApplicationTest(hooks);
  setupMirage(hooks);

  test('should list all widgets', async function(assert) {
      server.createList('widget', 3));
      await visit('/widgets');

      var widgets = this.element.querySelectorAll('.widget');
      assert.equal(widgets.length, 3);

      server.create('widget');

      widgets = this.element.querySelectorAll('.widget');
      assert.equal(widgets.length, 4); // this fails!
  });
});
```

**Comment by samselikoff** [comment]
@awrichar sorry you're running into trouble!

In your app when you call `DS.store.push(payload)`, how does your Ember app get the `payload`? Are you using web sockets? Making an ajax request?

`server.create()` will add data to Mirage's server but then your Ember app needs to do something to re-fetch that data.

**Comment by awrichar** [comment]
Thanks @samselikoff - in production, I get the payload via web sockets and then call `push()` to update Ember Data's local cache. My template uses `{{#each}}` to render the items, ie:

```
{{#each model as |widget|}}
<div class="widget"> ... </div>
{{/each}}
```

I don't have to make any special calls other than `push()` - as soon as the data is pushed to the store, Ember is smart enough to re-render the template.

Under test, I've tried using `server.create()` and directly using `push()` like so:

```
      this.owner.lookup('service:store').push({
          data: {id: '4', type: 'widget'},
      });
```

If I inspect the store after either method (`server.create()` or `store.push()`), it definitely contains 4 items. However, inspecting the DOM with `this.element.querySelectorAll()` only finds the original 3 items.

Basically I'm wondering if Ember/Mirage will re-render templates under test the same way it does in the "real world," and if so, whether there's a way to wait for that to happen.

**Comment by awrichar** [comment]
A few more data points:
* It occurred to me that maybe I could use [waitFor](https://github.com/emberjs/ember-test-helpers/blob/master/API.md#waitfor), but this appears to not be supported in acceptance tests. If I try to use it, I get `Uncaught Error: Must setup rendering context before attempting to interact with elements`.
* I also tried using [pauseTest](https://github.com/emberjs/ember-test-helpers/blob/master/API.md#pausetest) to visually inspect the HTML after creating the new item - it confirms the HTML element is not added no matter how long you wait.
* If I totally refresh the route with `await this.owner.lookup('route:' + currentRouteName()).refresh();`, then the new element shows up - but this is not simulating true app behavior (because in production I do not need to refresh in order to see the new element rendered).

**Comment by samselikoff** [comment]
I think `store.push` might be async – try `await`'ing it.

Another guess, try

```
run(() => {
  store.pushPayload(payload);
});
```

Also I would look into https://github.com/thoov/mock-socket to keep these tests as high-level as possible. Otherwise I would write a helper to do this so you can do something like

```js
let foo = server.create('foo');
simulateWebsocketEvent(server.serialize(foo));
```

**Comment by samselikoff** [comment]
Awesome! Glad you got unstuck.

You should definitely be able to wrap all that up into a helper that let's you do something like

```js
server.createList('widget', 3);
await visit('/widgets');
assert.equal(findAll('.widget').length, 3);

let newWidet = server.create('widget');
await receiveServerPayloadWith(newWidet);

assert.equal(findAll('.widget').length, 4);
```

As for #1273 the best person to ask would be @seansellek. You can definitely wire up websocket support with Mirage, it will just be a bit of manual wiring at the moment.

Going to close this issue for now, but feel free to keep the discussion going and let me know if you need any more help!
